lib/Sema/DerivedConformances.cpp

```diff
@@ -472,8 +470,19 @@ DerivedConformance::createBuiltinCall(ASTContext &ctx,
   ConcreteDeclRef declRef = decl;
   auto fnType = decl->getInterfaceType();
   if (auto genericFnType = fnType->getAs<GenericFunctionType>()) {
+    auto builtinModule = decl->getModuleContext();
     auto generics = genericFnType->getGenericSignature();
-    auto subs = SubstitutionMap::get(generics, typeArgs, conformances);
+
+    auto genericParams = generics.getGenericParams();
+    assert(genericParams.size() == typeArgs.size());
+
+    TypeSubstitutionMap map;
+    for (auto i : indices(genericParams))
+      map.insert({genericParams[i]->getAs<SubstitutableType>(), typeArgs[i]});
+
+    auto subs = SubstitutionMap::get(generics,
+                                     QueryTypeSubstitutionMap{map},
```

Comment:
A bunch of usages of QueryTypeSubstitutionMap could instead use an array of generic arguments in the correct order instead. If you want, you can add a new functor, like QueryReplacementTypeArray or something that does this. To map between from a generic parameter type to array index, you can use GenericParamKey:

```
  GenericParamKey key(paramTy);
  unsigned index = key.findIndexIn(genericParams);
```

Reply:
I tried this briefly but the functor only receives a SubstitutableType and it wasn't obvious how to derive the `GenericParamKey` for one of those. For now I'll leave this in as using `QueryTypeSubstitutionMap` but I agree with what you're getting at here... `QueryTypeSubstitutionMap` is a bit unwieldy and if you don't build it with a canonical type of the generic parameter as the key, you silently get no substitutions.

Reply:
OK I can do this clean-up in a follow-up PR along with simplifying CollectExistentialConformances.